### PR TITLE
Explicitly demarcate platform support

### PR DESCRIPTION
**Pull request by notgull**
As discussed in today's meeting, this commit creates two tiers of
support. Tier 1, which is what we actively test, and Tier 2, what we
hope compiles.

### CODE REVIEW

File: FEATURES.md [hunk 1]
```diff
@@ -230,6 +230,56 @@ Changes in the API that have been agreed upon but aren't implemented across all
...
+|64-bit x86 FreeBSD                  |`x86_64-unknown-freebsd`            |X11, Wayland   |
+|64-bit x86 NetBSD                   |`x86_64-unknown-netbsd`             |X11            |
+|32-bit x86 Linux with Musl          |`i686-unknown-linux-musl`           |X11, Wayland   |
+|64-bit RISC-V Linux with glibc      |`riscv64gc-unknown-linux-gnu`       |X11, Wayland   |
+|64-bit ARM Linux with glibc         |`aarch64-unknown-linux-gnu`         |X11, Wayland   |
+|64-bit ARM Linux with Musl          |`aarch64-unknown-linux-musl`        |X11, Wayland   |
+|64-bit PowerPC Linux with glibc     |`powerpc64le-unknown-linux-gnu`     |X11, Wayland   |
+|32-Bit ARM Linux with glibc         |`armv5te-unknown-linux-gnueabi`     |X11, Wayland   |
+|64-Bit Linux on IBM Supercomputers  |`s390x-unknown-linux-gnu`           |X11, Wayland   |
+|32-bit ARM Android                  |`arm-linux-androideabi`             |X11, Wayland   |
```
Comment (MarijnS95) [inline FEATURES.md:280]:
```suggestion
|32-bit ARM Android                  |`arm-linux-androideabi`             |Android        |
```

**Review by MarijnS95 — APPROVED** [review]

**Comment by kchibisov** [comment]
Should probably put it in a separate file, since we'll remove the `FEATURES.md` and then probably should also include it in the docs somehow?

**Comment by notgull** [comment]
> Should probably put it in a separate file, since we'll remove the `FEATURES.md` and then probably should also include it in the docs somehow?

Done and done

**Comment by notgull** [comment]
For the initial draft here I've gone with more platforms than we probably need. I think all the tier 1's are necessary, since we actively test them in CI and they directly correspond to the canonical versions of the `winit` backend's platforms.

For the tier 2's:

- ARM windows is becoming more popular, so we should make sure our bases are covered there.
- Do we want to support Windows 7? I added it here for discussion. I think we should, since lots of computers (especially old ones) still run Win7.
- Alpine Linux is sufficiently popular that I think we should be building `musl`. Same with FreeBSD. Not sure about NetBSD but issues have been filed on it in the past so I've added it.
- Other architecture Linux is fair game. I think we should be building ARM, I can take or leave the rest.

File: src/lib.rs [hunk 2]
```diff
@@ -147,6 +147,59 @@
 //! See the [`platform`] module for documentation on platform-specific cargo
 //! features.
 //!
+//! # Platform/Architecture Support
```
Comment (madsmtm) [inline src/lib.rs:150]: Either `#![doc = include_str!("PLATFORM_SUPPORT.md")]`, or to my preference, just avoid the `PLATFORM_SUPPORT.md` file altogether.

File: src/lib.rs [hunk 3]
```diff
@@ -147,6 +147,59 @@
...
+//! |32-Bit x86 Windows with glibc  |`i686-pc-windows-gnu`               |Win32          |
+//! |64-Bit x86 Windows with glibc  |`x86_64-pc-windows-gnu`             |Win32          |
+//! |32-Bit x86 Linux with glibc    |`i686-unknown-linux-gnu`            |X11, Wayland   |
+//! |64-Bit x86 Linux with glibc    |`x86_64-unknown-linux-gnu`          |X11, Wayland   |
+//! |64-Bit ARM Android             |`aarch64-linux-android`             |Android        |
+//! |64-Bit x86 Redox OS            |`x86_64-unknown-redox`              |Orbital        |
+//! |64-bit x64 macOS               |`x86_64-apple-darwin`               |AppKit         |
+//! |64-bit ARM macOS               |`aarch64-apple-darwin`              |AppKit         |
+//! |64-bit x64 iOS                 |`x86_64-apple-ios`                  |UIKit          |
+//! |64-bit ARM iOS                 |`aarch64-apple-ios`                 |UIKit          |
```
Comment (madsmtm) [inline src/lib.rs:177]: So, honestly, I feel like the statement "guaranteed to work" is too strong for the iOS backend. It doesn't really work properly (yet).

I'd be fine with calling it tier 2, and then I'll promote it myself once I feel like it's up there with the giants ;)

Same argument goes for Android, though it's not my call to make there.

Comment (madsmtm) [reply src/lib.rs:177]: Also, `aarch64-apple-ios-macabi` is also tested regularly (more regularly, actually).

**Review by madsmtm — APPROVED** [review]
Approved once the data is deduplicated somehow (see first comment)

File: src/lib.rs [hunk 4]
```diff
@@ -147,6 +147,60 @@
...
+//! |------------------------------------|------------------------------------|---------------|
+//! |64-Bit ARM Windows with MSVC        |`aarch64-pc-windows-msvc`           |Win32          |
+//! |32-Bit x86 Windows 7 with MSVC      |`i686-win7-windows-msvc`            |Win32          |
+//! |64-Bit x86 Windows 7 with MSVC      |`x86_64-win7-windows-msvc`          |Win32          |
+//! |64-bit x86 Linux with Musl          |`x86_64-unknown-linux-musl`         |X11, Wayland   |
+//! |64-bit x86 Linux with 32-bit glibc  |`x86_64-unknown-linux-gnux32`       |X11, Wayland   |
+//! |64-bit x86 Android                  |`x86_64-linux-android`              |Android        |
+//! |64-bit x64 iOS                      |`x86_64-apple-ios`                  |UIKit          |
+//! |64-bit ARM iOS                      |`aarch64-apple-ios`                 |UIKit          |
+//! |64-bit ARM iOS (macOS ABI)          |`aarch64-apple-ios-macabi`          |UIKit          |
```
Comment (madsmtm) [inline src/lib.rs:190]: Nit:

```suggestion
//! |64-bit ARM Mac Catalyst             |`aarch64-apple-ios-macabi`          |UIKit          |
```

File: src/lib.rs [hunk 5]
```diff
@@ -147,6 +147,60 @@
...
+//! |Target Name                    |Target Triple                       |APIs           |
+//! |-------------------------------|------------------------------------|---------------|
+//! |32-Bit x86 Windows with MSVC   |`i686-pc-windows-msvc`              |Win32          |
+//! |64-Bit x86 Windows with MSVC   |`x86_64-pc-windows-msvc`            |Win32          |
+//! |32-Bit x86 Windows with glibc  |`i686-pc-windows-gnu`               |Win32          |
+//! |64-Bit x86 Windows with glibc  |`x86_64-pc-windows-gnu`             |Win32          |
+//! |32-Bit x86 Linux with glibc    |`i686-unknown-linux-gnu`            |X11, Wayland   |
+//! |64-Bit x86 Linux with glibc    |`x86_64-unknown-linux-gnu`          |X11, Wayland   |
+//! |64-Bit ARM Android             |`aarch64-linux-android`             |Android        |
+//! |64-Bit x86 Redox OS            |`x86_64-unknown-redox`              |Orbital        |
```
Comment (jackpot51) [inline src/lib.rs:173]: I also test i686-unknown-redox and aarch64-unknown-redox regularly and they are functional. Should they be either Tier 1 or Tier 2?

Comment (notgull) [reply src/lib.rs:173]: The main idea behind this list is to determine which platforms should be tested in CI and which platforms should just be `cargo check`'d. Still, I'll take your word for it and add them to Tier 1.

**Review by jackpot51 — APPROVED** [review]

File: src/lib.rs [hunk 6]
```diff
@@ -147,6 +147,62 @@
...
+//! |32-Bit x86 Redox OS            |`i686-unknown-redox`                |Orbital        |
+//! |64-Bit ARM Redox OS            |`aarch64-unknown-redox`             |Orbital        |
+//! |64-bit x64 macOS               |`x86_64-apple-darwin`               |AppKit         |
+//! |64-bit ARM macOS               |`aarch64-apple-darwin`              |AppKit         |
+//! |32-bit WASM web browser        |`wasm32-unknown-unknown`            |DOM            |
```
Comment (daxpedda) [inline src/lib.rs:178]:
```suggestion
//! |32-bit Wasm web browser        |`wasm32-unknown-unknown`            |DOM            |
```

Comment (daxpedda) [inline src/lib.rs:178]:
```suggestion
//! |32-bit WASM web browser        |`wasm32-unknown-unknown`            |               |
```
I think its best to leave this empty ... "DOM" isn't appropriate here and I wouldn't know what other term to use in this case.

File: src/lib.rs [hunk 7]
```diff
@@ -147,6 +147,62 @@
...
 //! features.
 //!
+//! # Platform/Architecture Support
+//!
+//! Platform support on `winit` has two tiers: Tier 1 and Tier 2.
+//!
+//! - Tier 1 is **guaranteed to work**. Targets in this tier are actively tested both in CI and by
+//!   maintainers.
+//! - Tier 2 is **guaranteed to build**. Code compilation is tested in CI, but deeper testing is not
+//!   done.
```
Comment (daxpedda) [inline src/lib.rs:157]: I'm a bit confused, we don't actually test compilation of all these tier 2 targets in CI.
In the OP you said "we hope" compiles, which might be more appropriate here, but is probably very weird to add here.

**Review by daxpedda — CHANGES_REQUESTED** [review]
I think this requires a tier 3?
E.g. tier 2 is actually tested in CI, tier 3 is not but we believe it should work?
Unless we want to actually just go ahead and test them in CI (which I would be in favor of), then we should do this right here.

Or am I missing something?

File: src/lib.rs (hunk 6 above)
Comment (daxpedda) [reply src/lib.rs:178]: On second thought, I think this is the most appropriate here:
```suggestion
//! |32-bit WASM web browser        |`wasm32-unknown-unknown`            |`wasm-bindgen`|
```

File: src/lib.rs (hunk 7 above)
Comment (notgull) [reply src/lib.rs:157]: The goal is to start doing that in CI after we finalize this list.

Comment (daxpedda) [reply src/lib.rs:157]: Sure!

**Review by daxpedda — APPROVED** [review]
I went ahead and adjusted the wording around Web as I suggested.

**Pull request merged**